src/ZoneServer/World/Actors/Characters/Components/PersonalStorageComponent.cs

```diff
@@ -0,0 +1,512 @@
+			if (item.Amount <= amount)
+			{
+				amountRemoved = item.Amount;
+				_occupiedPositions -= 1;
+				_storageItems[position] = null;
+
+			}
+			// Remove item stacks
+			else
+			{
+				item.Amount -= amount;
+				amountRemoved = amount;
+			}
+			return amountRemoved;
+		}
+
+		/// <summary>
+		/// Adds amount to the item stack.
+		/// If item is not found in storage, returns 0.
+		/// If item is not stackable, returns 0.
+		/// </summary>
+		/// <param name="item">Item to add</param>
+		/// <param name="amount">Amount to stack to item</param>
+		/// <returns>Amount of stack added</returns>
+		private int AddItemStack(Item item, int amount)
+		{
+			if (!item.IsStackable)
+				return 0;
+
+			if (amount <= 0)
+				return 0;
+
+			if ((item.Amount + amount) < item.Data.MaxStack)
+			{
+				item.Amount += amount;
+				return amount;
+			}
+			else
+			{
+				var amountToAdd = item.Data.MaxStack - item.Amount;
+				item.Amount += amountToAdd;
+				return amountToAdd;
+			}
+		}
+
+		/// <summary>
+		/// Looks in storage for items that the given item
+		/// can stack to.
+		/// Does nothing if given item is not stackable.
+		/// </summary>
+		/// <param name="item">Item type look for</param>
+		/// <param name="foundPositions">Positions of found stackable items</param>
+		/// <param name="foundItems">Item objects of found items</param>
+		private void FindStackableItems(Item item, out List<int> foundPositions, out List<Item> foundItems)
+		{
+			foundPositions = new List<int>();
+			foundItems = new List<Item>();
+
+			// Stacks to existing items
+			if (item.IsStackable)
+			{
+				lock (_syncLock)
+				{
+					for (int i = 0; i < _maxStorage; i++)
+					{
+						if ((_storageItems[i] != null) && _storageItems[i].IsStackable && (_storageItems[i].Data.ClassName == item.Data.ClassName))
+						{
+							foundPositions.Add(i);
+							foundItems.Add(_storageItems[i]);
+						}
+					}
+				}
+			}
+		}
+
+		/// <summary>
+		/// Creates a new storage component
+		/// </summary>
+		/// <param name="account"></param>
+		public PersonalStorageComponent(Character character) : base(character)
+		{
+			_storageItems = new Item[_maxStorage];
+		}
+
+		/// <summary>
+		/// Adds an item to storage and removes from character's
+		/// inventory.
+		/// </summary>
+		/// <param name="worldId">Object id of the item</param>
+		/// <param name="amount">Amount to remove</param>
+		/// <returns></returns>
+		public StorageResult StoreItem(long worldId, int amount = 1)
+		{
+			var item = this.Character.Inventory.GetItem(worldId);
+			if (item == null)
+				return StorageResult.ItemNotFound;
+
+			Item itemToStore = new Item(item);
+			itemToStore.Amount = Math.Min(itemToStore.Amount, amount);
+
+			// For stackable items, finds existing storage items.
+			var existingPositions = new List<int>();
+			var existingItems = new List<Item>();
+			if (itemToStore.IsStackable)
+			{
+				this.FindStackableItems(itemToStore, out existingPositions, out existingItems);
```

Comment:
I opted to not take the FillStacks() approach. 

The reason for that is that I want to remove item from player **before** adding it to storage. It may be a bit redundant because I do check if item exists in player's inventory at very beggining of "StoreItem()" method but the "character.Inventory.Remove()" method I call has a return state and I want to make use of that and check it before proceeding further with adding the item to storage.